provider/dyn.go

```diff
@@ -0,0 +1,559 @@
+	// may be made configurable in the future but 20K records seems like enough for a few zones
+	cacheMaxSize = 20000
+)
+
+// A simple non-thread-safe cache with TTL. The TTL of the records is used here to
+// This cache is used to save on requests to DynAPI
+type cache struct {
+	contents map[string]*entry
+}
+
+type entry struct {
+	expires int64
+	ep      *endpoint.Endpoint
+}
+
+func (c *cache) Put(link string, ep *endpoint.Endpoint) {
+	// flush the whole cache on overflow
+	if len(c.contents) >= cacheMaxSize {
+		c.contents = make(map[string]*entry)
+	}
+
+	c.contents[link] = &entry{
+		ep:      ep,
+		expires: int64(time.Now().Unix()) + int64(ep.RecordTTL),
+	}
+}
+
+func (c *cache) Get(link string) *endpoint.Endpoint {
+	result, ok := c.contents[link]
+	if !ok {
+		return nil
+	}
+
+	now := int64(time.Now().Unix())
+
+	if result.expires < now {
+		delete(c.contents, link)
+		return nil
+	}
+
+	return result.ep
+}
+
+// DynConfig hold connection parameters to dyn.com and interanl state
+type DynConfig struct {
+	DomainFilter DomainFilter
+	ZoneIDFilter ZoneIDFilter
+	DryRun       bool
+	CustomerName string
+	Username     string
+	Password     string
+	AppVersion   string
+	DynVersion   string
+}
+
+// DynProvider is the actual interface impl.
+type dynProviderState struct {
+	DynConfig
+	Cache *cache
+}
+
+// ZoneChange is missing from dynect: https://help.dyn.com/get-zone-changeset-api/
+type ZoneChange struct {
+	ID     int              `json:"id"`
+	UserID int              `json:"user_id"`
+	Zone   string           `json:"zone"`
+	FQDN   string           `json:"FQDN"`
+	Serial int              `json:"serial"`
+	TTL    int              `json:"ttl"`
+	Type   string           `json:"rdata_type"`
+	RData  dynect.DataBlock `json:"rdata"`
+}
+
+// ZoneChangesResponse is missing from dynect: https://help.dyn.com/get-zone-changeset-api/
+type ZoneChangesResponse struct {
+	dynect.ResponseBlock
+	Data []ZoneChange `json:"data"`
+}
+
+// ZonePublishRequest is missing from dynect but the notes field is a nice place to let
+// external-dns report some internal info during commit
+type ZonePublishRequest struct {
+	Publish bool   `json:"publish"`
+	Notes   string `json:"notes"`
+}
+
+// ZonePublisResponse holds the status after publish
+type ZonePublisResponse struct {
+	dynect.ResponseBlock
+	Data map[string]interface{} `json:"data"`
+}
+
+// NewDynProvider initializes a new Dyn Provider.
+func NewDynProvider(config DynConfig) (Provider, error) {
+	return &dynProviderState{
+		DynConfig: config,
+		Cache: &cache{
+			contents: make(map[string]*entry),
+		},
+	}, nil
+}
+
+// filterAndFixLinks removes from `links` all the records we don't care about
+// and strops the /REST/ prefix
+func filterAndFixLinks(links []string, filter DomainFilter) []string {
+	var result []string
+	for _, link := range links {
+
+		// strip the /REST/ prefix as dynect always prepends it
+		link = strings.TrimPrefix(link, "/REST/")
```

Comment:
/REST/ can also be a global var if it's always the same, what do you think?

Reply:
Done